### PR TITLE
Mark skipped SLI tests as skipped, not success

### LINKED ISSUE #359: Mark skipped SLI tests as skipped, not success
Tests testing for/against crashing code using `crash_or_die`, `failbutnocrash_or_die` or `passorfailbutnocrash_or_die` can only be run if NEST has been compiled without MPI, because they `shpawn` a separate NEST instance.

This also applies to tests skipped by `exit_test_gracefully`.

Currently, these tests exit with exit code 0, flagging success when NEST is built with MPI. They should return a value marking that the test is skipped.

### PR DESCRIPTION
**Pull request by hakonsbm**
This PR fixes #359, and contains the following

- When NEST is compiled with MPI, tests that require NEST to be compiled without MPI are skipped with `exit_test_gracefully` instead of calling `quit_i` directly.
- Making `exit_test_gracefully` mark the test as skipped, not success.
- A new exit code, `200`, has been introduced to mark skipped tests.

### CODE REVIEW

File: testsuite/do_tests.sh.in [hunk 1]
```diff
@@ -412,6 +420,8 @@ echo >  "${TEST_LOGFILE}" "NEST v. @NEST_VERSION_VERSION@ testsuite log"
 echo >> "${TEST_LOGFILE}" "======================"
 echo >> "${TEST_LOGFILE}" "Running tests from ${TEST_BASEDIR}"
 
+CODES_SKIPPED=' 200 Skipped'
```
Comment (gtrensch) [inline testsuite/do_tests.sh.in]: I'm not sure about that but one may ask why these tests are skipped. `CODES_SKIPPED=' 200 Skipped (build with-mpi=OFF required)'` or something like that would make it clear. We have similar comments at other places, e.g. `Success (expected failure)`.

**Comment by gtrensch** [comment]
I tested the build with and without MPI. Everything works as expected. :+1: . I would approve this PR, but please see also my comments.

**Comment by hakonsbm** [comment]
I have now used named constants for the exit codes. Are they in the right place? And should I also add named constants for the other exit codes (scripterror, exception, etc)?

There are quite a few new changed files now, but most of the new changes are minor changes to tests. This is because in order to tell do_tests.sh why a test is skipped, I have introduced some new exit codes and a new SLI function, `exit_test_gracefully_reason[/stringtype]`. This way do_tests.sh can differentiate different reasons for skipping, and give associated comments.

**Comment by gtrensch** [comment]
@hakonsbm many thanks for the work ! I think the constants are in the right place. Replacing all the other exit codes would of course be a good idea. But I'm not sure if this PR is the right place to do this refactoring task.

File: sli/slistartup.h [hunk 2]
```diff
@@ -40,6 +40,15 @@
 #include "slimodule.h"
 #include "token.h"
 
+// Exit codes
+#define NEST_EXITCODE_SKIPPED 200
+#define NEST_EXITCODE_SKIPPED_NO_MPI 201
+#define NEST_EXITCODE_SKIPPED_HAVE_MPI 202
+#define NEST_EXITCODE_SKIPPED_NO_OPENMP 203
+#define NEST_EXITCODE_SKIPPED_NO_GSL 204
+#define NEST_EXITCODE_SKIPPED_NO_MUSIC 205
```
Comment (heplesser) [inline sli/slistartup.h]: Add the other numeric values from `slistartup.cpp` here (in increasing order). These exit codes are not related to NEST as such, so drop `NEST_` from the names. Maybe add a comment that the exit code range 200-215 is reserved for test skipping exits, and point out that any new skipping codes must be added to `testsuite/do_tests.sh.in`.

File: testsuite/mpitests/issue-600.sli [hunk 3]
```diff
@@ -40,13 +40,8 @@
...
 % this test should only run if we have GSL
-is_threaded not 
-statusdict/have_gsl :: not
-or 
-{ 
-  exit_test_gracefully
-} 
-if
+is_threaded not {(no_openmp) exit_test_gracefully_reason} if
+statusdict/have_gsl :: not {(no_gsl) exit_test_gracefully_reason} if
```
Comment (heplesser) [inline testsuite/mpitests/issue-600.sli]: We have a small number of skipping conditions which lead to a lot of repeated code. I think we could make things much tidier by introducing functions for each of the cases, e.g.,
```
/skip_if_not_threaded { is_threaded not { /no_openmp exit_test_gracefully } if } def
```
Then, in this case, the test code would just read
```
skip_if_not_threaded
skip_if_without_gsl
```

File: lib/sli/unittest.sli [hunk 4]
```diff
@@ -47,9 +47,19 @@ is_threaded_with_openmp not { exit_test_gracefully } if
...
+/exit_test_gracefully_reason[/stringtype]
+{
+  dup (no_mpi) eq {statusdict/exitcodes/skipped_no_mpi :: quit_i} if
+  dup (have_mpi) eq {statusdict/exitcodes/skipped_have_mpi :: quit_i} if
+  dup (no_openmp) eq {statusdict/exitcodes/skipped_no_openmp :: quit_i} if
+  dup (no_gsl) eq {statusdict/exitcodes/skipped_no_gsl :: quit_i} if
+  dup (no_music) eq {statusdict/exitcodes/skipped_no_music :: quit_i} if
+  statusdict/exitcodes/scripterror :: quit_i
+} bind def
+
```
Comment (heplesser) [inline lib/sli/unittest.sli]: I would also change all "openmp" into "no_threads" or "not_threaded". OpenMP is an implementation detail.

File: testsuite/do_tests.sh.in [hunk 5]
```diff
@@ -249,7 +249,8 @@ run_test ()
 
     param_script="$1"
     param_success="$2"
-    param_failure="$3"
+    param_skipped="$3"
+    param_failure="$4"
```
Comment (heplesser) [inline testsuite/do_tests.sh.in]: Please update the documentation of `run_test()` above to include `param_skipped`.

File: testsuite/do_tests.sh.in [hunk 6]
```diff
@@ -288,11 +289,18 @@ run_test ()
...
     msg_clean=${msg_dirty%%,*}
     if test "${msg_dirty}" != "${param_success}" ; then
         TEST_PASSED=$(( ${TEST_PASSED} + 1 ))
         explanation="${msg_clean}"
         junit_failure=
+    elif test "${msg_dirty_skip}" != "${param_skipped}" ; then
+        TEST_PASSED=$(( ${TEST_PASSED} + 1 ))
+        msg_dirty=${param_skipped##* ${exit_code} }
+        msg_clean=${msg_dirty%%,*}
+        explanation="${msg_clean}"
```
Comment (heplesser) [inline testsuite/do_tests.sh.in]: If I understand this code right, then this section handles skipped tests. The Travis log shows that correct messages are shown for individual skipped tests, but the skipped tests should be counted as skipped, not as passed, and the final summary should show them as skipped.

File: testsuite/do_tests.sh.in [hunk 7]
```diff
@@ -288,11 +289,18 @@ run_test ()
...
     msg_dirty=${param_success##* ${exit_code} }
+    msg_dirty_skip=${param_skipped##* ${exit_code} }
     msg_clean=${msg_dirty%%,*}
     if test "${msg_dirty}" != "${param_success}" ; then
         TEST_PASSED=$(( ${TEST_PASSED} + 1 ))
         explanation="${msg_clean}"
         junit_failure=
+    elif test "${msg_dirty_skip}" != "${param_skipped}" ; then
+        TEST_PASSED=$(( ${TEST_PASSED} + 1 ))
+        msg_dirty=${param_skipped##* ${exit_code} }
```
Comment (heplesser) [inline testsuite/do_tests.sh.in]: The right hand side here is the same as on line 292, maybe no need to repeat?
